File: ios/Packages/AppFeatures/Sources/TasbeehFeature/TasbeehModels.swift
import Foundation

/// Bundled dhikr presets (docs/features/tasbeeh.md). A backend-driven list is
/// a natural future extension (mirrors wird_templates); not required for M2.
public struct DhikrPreset: Identifiable, Equatable, Sendable {
    public let id: String
    public let arabicText: String

    public init(id: String, arabicText: String) {
        self.id = id
        self.arabicText = arabicText
    }

    /// Sentinel selection meaning "let the user type custom text".
    public static let custom = DhikrPreset(id: "custom", arabicText: "")

    public static let bundled: [DhikrPreset] = [
        DhikrPreset(id: "subhanallah", arabicText: "سبحان الله"),
        DhikrPreset(id: "alhamdulillah", arabicText: "الحمد لله"),
        DhikrPreset(id: "allahu_akbar", arabicText: "الله أكبر"),
        DhikrPreset(id: "la_ilaha_illallah", arabicText: "لا إله إلا الله"),
        DhikrPreset(id: "subhanallahi_wabihamdihi", arabicText: "سبحان الله وبحمده"),
        DhikrPreset(id: "astaghfirullah", arabicText: "أستغفر الله"),
    ]

    public static let commonTargets = [33, 99, 100, 1000]
}

/// A completed counting set, persisted to history.
public struct TasbeehHistoryEntry: Codable, Equatable, Identifiable, Sendable {
    public let id: UUID
    public let presetId: String?
    public let customText: String?
    public let target: Int
    public let actualCount: Int
    public let completedAt: Date

    public init(
        id: UUID = UUID(),
        presetId: String?,
        customText: String?,
        target: Int,
        actualCount: Int,
        completedAt: Date
    ) {
        self.id = id
        self.presetId = presetId
        self.customText = customText
        self.target = target
        self.actualCount = actualCount
        self.completedAt = completedAt
    }
}

public struct TasbeehStats: Equatable, Sendable {
    public let totalCount: Int
    public let setsCompleted: Int

    public init(totalCount: Int, setsCompleted: Int) {
        self.totalCount = totalCount
        self.setsCompleted = setsCompleted
    }

    public static func from(history: [TasbeehHistoryEntry]) -> TasbeehStats {
        TasbeehStats(
            totalCount: history.reduce(0) { $0 + $1.actualCount },
            setsCompleted: history.count
        )
    }
}

/// Persistence boundary for history (mirrors ConfigKit.ConfigStoring).
public protocol TasbeehHistoryStoring: Sendable {
    func load() -> [TasbeehHistoryEntry]
    func save(_ history: [TasbeehHistoryEntry])
}

public struct FileTasbeehHistoryStore: TasbeehHistoryStoring {
    private let fileURL: URL

    public init(directory: URL) {
        self.fileURL = directory.appendingPathComponent("tasbeeh-history.json")
    }

    public func load() -> [TasbeehHistoryEntry] {
        guard let data = try? Data(contentsOf: fileURL) else { return [] }
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return (try? decoder.decode([TasbeehHistoryEntry].self, from: data)) ?? []
    }

    public func save(_ history: [TasbeehHistoryEntry]) {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        guard let data = try? encoder.encode(history) else { return }
        try? FileManager.default.createDirectory(
            at: fileURL.deletingLastPathComponent(), withIntermediateDirectories: true
        )
        try? data.write(to: fileURL, options: .atomic)
    }
}

File: android/feature/tasbeeh/src/main/kotlin/com/fatwabot/feature/tasbeeh/TasbeehModels.kt
package com.fatwabot.feature.tasbeeh

import com.fatwabot.core.common.HapticsProviding
import java.io.File
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json

/** Mirror of iOS DhikrPreset (docs/features/tasbeeh.md). */
data class DhikrPreset(val id: String, val arabicText: String) {
    companion object {
        val CUSTOM = DhikrPreset("custom", "")

        val bundled = listOf(
            DhikrPreset("subhanallah", "سبحان الله"),
            DhikrPreset("alhamdulillah", "الحمد لله"),
            DhikrPreset("allahu_akbar", "الله أكبر"),
            DhikrPreset("la_ilaha_illallah", "لا إله إلا الله"),
            DhikrPreset("subhanallahi_wabihamdihi", "سبحان الله وبحمده"),
            DhikrPreset("astaghfirullah", "أستغفر الله"),
        )

        val commonTargets = listOf(33, 99, 100, 1000)
    }
}

@Serializable
data class TasbeehHistoryEntry(
    val id: String,
    val presetId: String?,
    val customText: String?,
    val target: Int,
    val actualCount: Int,
    val completedAtEpochSeconds: Long,
)

data class TasbeehStats(val totalCount: Int, val setsCompleted: Int) {
    companion object {
        fun from(history: List<TasbeehHistoryEntry>) = TasbeehStats(
            totalCount = history.sumOf { it.actualCount },
            setsCompleted = history.size,
        )
    }
}

/** Persistence boundary (mirrors core:config's FileConfigStore pattern). */
interface TasbeehHistoryStoring {
    fun load(): List<TasbeehHistoryEntry>
    fun save(history: List<TasbeehHistoryEntry>)
}

class FileTasbeehHistoryStore(private val file: File) : TasbeehHistoryStoring {
    private val json = Json { ignoreUnknownKeys = true }

    private val serializer = ListSerializer(TasbeehHistoryEntry.serializer())

    override fun load(): List<TasbeehHistoryEntry> = runCatching {
        if (!file.exists()) emptyList() else json.decodeFromString(serializer, file.readText())
    }.getOrDefault(emptyList())

    override fun save(history: List<TasbeehHistoryEntry>) {
        runCatching {
            file.parentFile?.mkdirs()
            val tmp = File(file.parentFile, "${file.name}.tmp")
            tmp.writeText(json.encodeToString(serializer, history))
            tmp.renameTo(file)
        }
    }
}
